### PR TITLE
Revert improved filtering for variations

### PR DESCRIPTION
**Pull request by Konamiman**
### All Submissions:

* [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This pull request reverts the functionality introduced in #26260 (later refined by #27175, #27190, #27508) in which filtering by
attribute using the layered nav widget was improved to handle the cases of variations out of stock. The revert is a response to the numerous problems related to filtering reported by users in WooCommerce 4.4 and 4.5, including two more that we hadn't detected previously:

* Attributes not used for variations aren't handled correctly.
* Products marked as "hidden for catalog" are still being counted.

We'll figure out a better way to achieve the same functionality and re-implement the feature in a future version of WooCommerce.

Not all the code has been reverted, only the code that resulted in visible functionality changes. Thus, the code that generates
term relationships for variations is still in place to keep database consistency and to keep the reverting changes to the minimum needed.

May close (hopefully) https://github.com/woocommerce/woocommerce/issues/27419.

**Note for reviewers:** To check the code changes in the widget itself you may find it useful to open [the diff of this PR against 4.3.0](https://github.com/woocommerce/woocommerce/compare/4.3.0...revert-improved-filtering-for-variations), go to "Files changed" and search for `includes/widgets/class-wc-widget-layered-nav.php` (I didn't fully revert this one since there are small improvements that can be kept, e.g. for the code sniffer).

### How to test the changes in this Pull Request:

Basically, verify that the filtering widget works as in WooCommerce 4.3. This implies that even if "hide out of stock items" is set, there are simple and variable products, and at least one of them has stock, the widget will count all the attributes including:

* The ones used by simple products.
* The ones used by variable products, but not for variations.
* The ones used by variable products, for variations, even those not having stock.

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Fix - Revert the changes in filtering by attribute that were introduced in WooCommerce 4.4

### CODE REVIEW

**Review by ObliviousHarmony — APPROVED** [review]
LGTM!

File: templates/content-product.php [hunk 1]
```diff
@@ -20,7 +20,7 @@
 global $product;
 
 // Ensure visibility.
-if ( empty( $product ) || false === wc_get_loop_product_visibility( $product->get_id() ) || ! $product->is_visible() ) {
```
Comment (claudiosanches) [inline templates/content-product.php:23]: We should change the `wc_get_loop_product_visibility()` function to make sure it doesn't runs `wc_set_loop_product_visibility()` anymore.
Note that some people may have replaced those templates in their themes.

Comment (Konamiman) [reply templates/content-product.php:23]: Not sure if I understand what you mean. `wc_get_loop_product_visibility` is just doing a call to `wc_get_loop_prop`, or do you mean something else?

**Comment by vedanshujain** [comment]
I am restoring all the related branches for review, will delete them later myself (so please don't delete them now :) ).

File: includes/widgets/class-wc-widget-layered-nav.php [hunk 2]
```diff
@@ -344,93 +344,51 @@ protected function layered_nav_dropdown( $terms, $taxonomy, $query_type ) {
...
-		$main_tax_query_sql = $this->convert_tax_query_to_sql( $main_tax_query );
-		$term_ids_sql       = '(' . implode( ',', array_map( 'absint', $term_ids ) ) . ')';
+		$meta_query     = new WP_Meta_Query( $meta_query );
+		$tax_query      = new WP_Tax_Query( $tax_query );
+		$meta_query_sql = $meta_query->get_sql( 'post', $wpdb->posts, 'ID' );
+		$tax_query_sql  = $tax_query->get_sql( $wpdb->posts, 'ID' );
+		$term_ids_sql   = '(' . implode( ',', array_map( 'absint', $term_ids ) ) . ')';
 
 		// Generate the first part of the query.
 		// This one will return non-variable products and variable products with concrete values for the attributes.
```
Comment (vedanshujain) [inline includes/widgets/class-wc-widget-layered-nav.php:365]: no need for this comment anymore

**Review by vedanshujain — CHANGES_REQUESTED** [review]
Great work! I have added some comments inline, and some suggestions in the summary comments (which are optional).

Improvements we should do soon (these would have been blockers if we were not doing it in a fix release)

- [ ] Update function `wc_update_440_insert_attribute_terms_for_variable_products` does too much for running in a 30s cron. The time it will take is mostly dependent on how many variations are there for products on the page. We should instead loop on variations (as opposed to their parent product), and instead of the page, we should fetch on ID and specify explicit order. Fetching using the page is prone to race conditions and filters interfering with the query so we should make it as explicit as possible.

Suggestions (not blockers, can be done in a different PR) 
- [ ] When [deleting associated taxanomies](https://github.com/woocommerce/woocommerce/blob/d48f1d4e2efcfab8206799df2085f021ec273722/includes/data-stores/class-wc-product-variable-data-store-cpt.php#L664), fetch the term_ids directly and use `wp_remove_object_terms`. Right now we are effectively looping across all the product attributes which is going to be very slow for large shops.
- [ ] Same with the block add to [sync attribute to terms table](https://github.com/woocommerce/woocommerce/blob/d48f1d4e2efcfab8206799df2085f021ec273722/includes/data-stores/class-wc-product-variation-data-store-cpt.php#L499) as well. We should find a way to avoid calling `wc_get_attribute_taxonomy_names` and instead fetch the existing attributes and delete them directly.
- [ ] Same in [product variation delete](https://github.com/woocommerce/woocommerce/blob/d48f1d4e2efcfab8206799df2085f021ec273722/includes/class-wc-product-variation.php#L601).

**Comment by Konamiman** [comment]
> Update function wc_update_440_insert_attribute_terms_for_variable_products does too much for running in a 30s cron.

This migration is not needed once the revert is in place, since these extra entries in the term relationships table won't be used for anything. I didn't remove it to keep the revert as simple as possible, but we'll probably be able to remove this migration (and the already existing entries) later.

And the same for the other suggestions about syncing variation terms.

File: includes/class-wc-query.php [hunk 3]
```diff
@@ -395,52 +382,6 @@ public function remove_product_query_filters( $posts ) {
...
-	 *
-	 * We also cache the post visibility so that it isn't checked again when displaying the posts list.
-	 *
-	 * @since 4.4.0
-	 * @param int      $count Original posts count, as supplied by the found_posts filter.
-	 * @param WP_Query $query The current WP_Query object.
-	 *
-	 * @return int Adjusted posts count.
-	 */
-	public function adjust_posts_count( $count, $query ) {
```
[the remark this answers is not in the document]
Comment (Konamiman) [reply includes/class-wc-query.php:412]: Ok. Note however that we'll probably need to reintroduce the same function later when we re-implement the thing, so maybe it's better to just make it return the count unchanged as you say, but without deprecating it?

**Review by vedanshujain — APPROVED** [review]
Approving the PR with pending improvements in order to keep the changeset as small as possible for the fix release.

**Pull request merged**